crates/tauri/src/tray/mod.rs

```diff
@@ -342,10 +342,29 @@ impl<R: Runtime> TrayIconBuilder<R> {
   /// Builds and adds a new [`TrayIcon`] to the system tray.
   pub fn build<M: Manager<R>>(self, manager: &M) -> crate::Result<TrayIcon<R>> {
     let id = self.id().clone();
-    let inner = self.inner.build()?;
+
+    struct SafeSend<T>(T);
+    unsafe impl<T> Send for SafeSend<T> {}
+
+    // SAFETY:
+    // the menu within this builder was created on main thread
+    // and will be accessed on the main thread
+    let inner = &mut Some(self.inner) as *mut _ as isize;
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    let inner = manager
+      .app_handle()
+      .run_on_main_thread(move || {
+        let inner = unsafe { &mut *(inner as *mut Option<tray_icon::TrayIconBuilder>) };
+
+        // SAFETY: will only be accessed on main thread
+        let _ = tx.send(inner.take().unwrap().build().map(SafeSend));
+      })
+      .and_then(|_| rx.recv().map_err(|_| crate::Error::FailedToReceiveMessage))??;
```

Comment:
The cast just feels wrong to me, maybe something like this?

```suggestion
    struct SafeSendTrayIconBuilder(tray_icon::TrayIconBuilder);
    unsafe impl Send for SafeSendTrayIconBuilder {}
    impl SafeSendTrayIconBuilder {
      fn build(self) -> Result<tray_icon::TrayIcon, tray_icon::Error> {
        self.0.build()
      }
    }

    // SAFETY:
    // the menu within this builder was created on main thread
    // and will be accessed on the main thread
    let inner = SafeSendTrayIconBuilder(self.inner);

    let (tx, rx) = std::sync::mpsc::channel();
    let inner = manager
      .app_handle()
      .run_on_main_thread(move || {
        // SAFETY: will only be accessed on main thread
        let _ = tx.send(inner.build().map(SafeSend));
      })
      .and_then(|_| rx.recv().map_err(|_| crate::Error::FailedToReceiveMessage))??;
```

Reply:
nice one, I tried similar thing but it didn't work, looks like the trick is moving `.build()` into a method on the new type, otherwise the closure captures the `dyn ContextMenu` and then rust complains

Reply:
Found that we don't event need to add a new `build` method, we can just access the inner value through a method and all works
```rs
struct UnsafeSend<T>(T);
unsafe impl<T> Send for UnsafeSend<T> {}

impl<T> UnsafeSend<T> {
  fn take(self) -> T {
    self.0
  }
}

```